### PR TITLE
Netty4.1: remove our handler when original handler is removed

### PR DESCRIPTION
**Pull request by laurit**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1373

### CODE REVIEW

File: instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyChannelPipelineInstrumentation.java [hunk 1]
```diff
@@ -106,39 +119,78 @@ public static void addHandler(
...
+      }
+    }
+  }
+
+  public static class ChannelPipelineRemoveAdvice {
+    @Advice.OnMethodEnter(suppress = Throwable.class)
+    public static void removeHandler(
+        @Advice.This ChannelPipeline pipeline, @Advice.Argument(0) ChannelHandler handler) {
+      ChannelHandler ourHandler =
+          InstrumentationContext.get(ChannelHandler.class, ChannelHandler.class).get(handler);
```
Comment (mateuszrzeszutek) [inline instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyChannelPipelineInstrumentation.java:156]: Shouldn't we also remove the handler from the `ContextStore`?

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyChannelPipelineInstrumentation.java (hunk 1 above)
Comment (laurit) [reply instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/NettyChannelPipelineInstrumentation.java]: Thanks, removed it.

**Review by trask — COMMENTED** [review]
can you add a test, similar to the one for removing future listeners? https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/netty/netty-4.1/javaagent/src/test/groovy/ChannelFutureTest.groovy

File: instrumentation/netty/netty-4.1/javaagent/src/test/groovy/ChannelPipelineTest.groovy [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+import io.netty.channel.DefaultChannelPipeline
+import io.netty.channel.embedded.EmbeddedChannel
+import io.netty.handler.codec.http.HttpClientCodec
+import io.opentelemetry.instrumentation.test.AgentInstrumentationSpecification
+import io.opentelemetry.javaagent.instrumentation.netty.v4_1.client.HttpClientTracingHandler
+
+class ChannelPipelineTest extends AgentInstrumentationSpecification {
+  // regression test for https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1373
+  def "test remove our handler #testName"() {
```
Comment (mateuszrzeszutek) [inline instrumentation/netty/netty-4.1/javaagent/src/test/groovy/ChannelPipelineTest.groovy:14]:
```suggestion
  @Unroll
  def "test remove our handler #testName"() {
```

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelInstrumentation.java [hunk 3]
```diff
@@ -40,7 +39,7 @@
   @Override
   public void transform(TypeTransformer transformer) {
     transformer.applyAdviceToMethod(
-        isMethod().and(nameStartsWith("write")),
+        isMethod().and(named("write").or(named("writeAndFlush"))),
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelInstrumentation.java:42]: epoll & kqueue channel implementations have methods starting with write that shouldn't be instrumented

**Pull request merged**